tests/test_functional.py

```diff
@@ -0,0 +1,200 @@
+import pytest
+import obal
+
+
+FIXTURE_DIR = os.path.join(
+    os.path.dirname(os.path.realpath(__file__)),
+    'fixtures',
+)
+
+TESTREPO_DIR = os.path.join(FIXTURE_DIR, 'testrepo')
+MOCKBIN_DIR = os.path.join(FIXTURE_DIR, 'mockbin')
+
+DEFAULT_ARGS = []
+if os.environ.get('TRAVIS', None):
+    DEFAULT_ARGS.extend(['-e', 'ansible_remote_tmp=/tmp/ansible-remote'])
+
+
+def obal_cli_test(func=None, repotype='upstream'):
+    if func is None:
+        return functools.partial(obal_cli_test, repotype=repotype)
+
+    @functools.wraps(func)
+    def func_wrapper(*args, **kwargs):
+        tempdir = tempfile.mkdtemp()
+        repodir = os.path.join(tempdir, 'repo')
+        shutil.copytree(TESTREPO_DIR, repodir)
+
+        oldcwd = os.getcwd()
+        oldpath = os.environ['PATH']
+        os.chdir(os.path.join(repodir, repotype))
+        os.environ['PATH'] = "{}:{}".format(MOCKBIN_DIR, oldpath)
+        os.environ['MOCKBIN_LOG'] = os.path.join(tempdir, 'mockbin.log')
+
+        subprocess.call(['git', 'init'])
+
+        func(*args, **kwargs)
+
+        os.chdir(oldcwd)
+        os.environ['PATH'] = oldpath
+        os.environ.pop('MOCKBIN_LOG')
+
+        shutil.rmtree(tempdir, ignore_errors=True)
+
+    return func_wrapper
+
+
+def assert_mockbin_log(content):
+    expected_log = "\n".join(content)
+    expected_log = expected_log.replace('{bin}', MOCKBIN_DIR)
+    with open(os.environ['MOCKBIN_LOG']) as mockbinlog:
+        log = mockbinlog.read().strip()
+        assert log == expected_log
+
+
+def setup_upstream(upstream_path):
+    # create a cloneable upstream repo, what we ship in fixtures is not
+    subprocess.call(['git', 'init'], cwd=upstream_path)
+    subprocess.call(['git', 'annex', 'init'], cwd=upstream_path)
+    subprocess.call(['git', 'add', '.'], cwd=upstream_path)
+    subprocess.call(['git', 'annex', 'addurl', '--file',
+                     'packages/hello/hello-2.10.tar.gz',
+                     'http://ftp.gnu.org/gnu/hello/hello-2.10.tar.gz'],
+                    cwd=upstream_path)
+    subprocess.call(['git', 'commit', '-a', '-m', 'init'], cwd=upstream_path)
+
+
+@obal_cli_test(repotype='upstream')
+def test_obal_noargs():
+    sys.argv = ['obal'] + DEFAULT_ARGS
+    with pytest.raises(SystemExit) as excinfo:
+        obal.main()
+    assert excinfo.value.code == 2
+
+
+@obal_cli_test(repotype='upstream')
+def test_obal_scratch_upstream_hello():
+    sys.argv = ['obal', 'scratch', 'hello'] + DEFAULT_ARGS
+    with pytest.raises(SystemExit) as excinfo:
+        obal.main()
+
+    assert excinfo.value.code == 0
+    assert os.path.exists('packages/hello/hello-2.10.tar.gz')
+
+    expected_log = [
+        "['{bin}/tito', 'release', '--scratch', 'dist-git', '-y']"
+    ]
+    assert_mockbin_log(expected_log)
+
+
+@obal_cli_test(repotype='upstream')
+def test_obal_release_upstream_hello():
+    sys.argv = ['obal', 'release', 'hello'] + DEFAULT_ARGS
+    with pytest.raises(SystemExit) as excinfo:
+        obal.main()
```

Comment:
The lines 98-100 here appear to be a pattern throughout the tests. Something we could DRY up without harming the readability of the tests?

Reply:
Would the assert need to be in the with block since it references `excinfo` ?

I could see it being nice from a testing perspective to either:

 1) Split the running of the command from the assertion of success or failure of the command
 2) Have two methods that call similar code but expect success or failure

I am thinking of it as being able to run:

```
assert_obal_success 'release hello'
assert_obal_failure 'release blah-blah not real command'
```

My reasoning is when looking at testing, I want to know I am performing an assertion vs. running helper code. This I find helps me with readability of tests.

Reply:
I am not sure. Will test it out.

Reply:
Tried to drain the swamp a bit, added `assert_obal_success` and `assert_obal_failure` wrappers, based on the above `run_obal`. And yes, the `excinfo` assertion has to happen in the same func as the `with`.